### ISSUE TITLE
NIOFoundationCompat+ByteBuffer: Support JSONSerialization directly

### ISSUE THREAD

**Issue by weissi** [apple/swift-nio#1891]
In `NIOFoundationCompat` we already have a bunch of useful convenience for `ByteBuffer` + `JSONEncoder`/`JSONDecoder`. We should have the same for `JSONSerialization`. For example:

    try JSONSerialization.jsonObject(with: Data(buffer: byteBuffer), ...)

is tedious, we could directly have

    try JSONSerialization.jsonObject(buffer: byteBuffer, ...)

**Comment by agnosticdev** [comment]
Wanted to get your thoughts on the following:
```swift

extension ByteBuffer {
    
    /// Attempts to derive a Foundation object from a ByteBuffer and return it  as `T`.
    ///
    /// - parameters:
    ///    - type: The Founation type that is attempting to be derived from the ByteBuffer.
    /// - returns: The Foundation value if successful or `nil` if there was an issue creating the Foundation type.
    @inlinable
    public func object<T>(_ type: T.Type) throws -> T? {
        guard let bytes = self.getData(at: 0, length: self.readableBytes),
              let t = try JSONSerialization.jsonObject(with: bytes, options: .mutableContainers) as? T else {
            return nil
        }
        return t
    }
    
    /// Attempts to derive a Foundation object from a ByteBuffer that is transformed into a InputStream and returned  as `T`.
    ///
    /// - parameters:
    ///    - type: The Founation type that is attempting to be derived from the ByteBuffer.
    /// - returns: The Foundation value if successful or `nil` if there was an issue creating the Foundation type.
    @inlinable
    public func objectFromStream<T>(_ type: T.Type) throws -> T? {
        guard let bytes = self.getData(at: 0, length: self.readableBytes) else {
            return nil
        }
        let stream = InputStream(data: bytes)
        stream.open()
        guard let t = try JSONSerialization.jsonObject(with: stream, options: .mutableContainers) as? T else {
            return nil
        }
        stream.close()
        return t
    }
}
```
With the usage looking something like the following:
```swift

let array = ["String1", "String2", "String3"]
let dictionary = ["key1": "val1", "key2": "val2", "key3": "val3"]

var dataArray = try JSONSerialization.data(withJSONObject: self.array, options: .prettyPrinted)
var dataDictionary = try JSONSerialization.data(withJSONObject: self.dictionary, options: .prettyPrinted)

let arrayDispatchData = dataArray.withUnsafeBytes { DispatchData(bytes: $0) }
let dictDispatchData = dataDictionary.withUnsafeBytes { DispatchData(bytes: $0) }

let arrayByteBuffer = ByteBuffer(dispatchData: arrayDispatchData)
let dictByteBuffer = ByteBuffer(dispatchData: dictDispatchData)

var foundationArray: [String] = []
var foundationDict: [String: String] = [:]
XCTAssertNoThrow(foundationArray = try arrayByteBuffer.object(Array<String>.self)!)
XCTAssertEqual(foundationArray, self.array)

XCTAssertNoThrow(foundationDict = try dictByteBuffer.object(Dictionary<String, String>.self)!)
XCTAssertEqual(foundationDict, self.dictionary)

```

Is this what you had in mind or were you looking for something different?
If this is what you had in mind, did you also see a need to include the `InputStream` API as well?

**Comment by weissi** [comment]
@agnosticdev Thanks, pretty close! I'd do `extension JSONSerialization` and add a `jsonObject` method that is the same signature as `JSONSerialization.jsonObject(with:Data, ...)` except that it has `buffer: ByteBuffer`.

This is incorrect btw: `let bytes = self.getData(at: 0, length: self.readableBytes)` and should be `Data(buffer: self)`.

And we don't need the `objectFromStream` method because we already have the bytes, so we don't need to make it into a stream, then coalesce everything and go through the other code path essentially.

**Comment by agnosticdev** [comment]
Going the extension route for `JSONSerialization`:

```swift
extension JSONSerialization {
    
    /// Attempts to derive a Foundation object from a ByteBuffer and return it  as `T`.
    ///
    /// - parameters:
    ///    - type: The Founation type that is attempting to be derived from the ByteBuffer.
    ///    - buffer: The ByteBuffer being used to derive the Foundation type.
    /// - returns: The Foundation value if successful or `nil` if there was an issue creating the Foundation type.
    @inlinable
    public func jsonObject<T>(_ type: T.Type, buffer: ByteBuffer) throws -> T? {
        guard let t = try JSONSerialization.jsonObject(with: Data(buffer: buffer), options: .mutableContainers) as? T else {
            return nil
        }
        return t
    }
}

```

Ends up running into two different errors when attempting to use/test the new API.  First when using a static version of `JSONSerialization`, I see:

```
Instance member 'jsonObject' cannot be used on type 'JSONSerialization'; did you mean to use a value of this type instead?
XCTAssertNoThrow(foundationArray = try JSONSerialization.jsonObject(Array<String>.self, buffer: arrayByteBuffer)!)

```

Okay, makes sense, but then when I created an instance of `JSONSerialization` it also resulted in an error:

```
let serializer = JSONSerialization()
...
JSONSerialization+ByteBufferTest.swift:29: error: -[NIOFoundationCompatTests.JSONSerializationByteBufferTest testSerializationRoundTripTest] : *** +[NSJSONSerialization allocWithZone:]: Do not create instances of NSJSONSerialization in this release (NSInvalidArgumentException)

```
Let me know your thoughts.

**Comment by weissi** [comment]
> Instance member 'jsonObject' cannot be used on type 'JSONSerialization'; did you mean to use a value of this type instead?

You need to make `jsonObject` a `public static func` not a `public func`, then it should work.

**Comment by agnosticdev** [comment]
Thanks @weissi!  That did work.

**Comment by agnosticdev** [comment]
Looks like this one can be closed now that #1906 is merged?

**Comment by Lukasa** [comment]
Agreed.
